salt/transport/rabbitmq.py

```diff
@@ -0,0 +1,1243 @@
+    def __init__(self, opts, **kwargs):
+        pass
+
+    # an init for the singleton instance to call
+    def __singleton_init__(self, opts, **kwargs):
+        self.opts = dict(opts)
+        self.ttype = "rabbitmq"
+
+        # crypt defaults to 'aes'
+        self.crypt = kwargs.get("crypt", "aes")
+
+        if "master_uri" in kwargs:
+            self.opts["master_uri"] = kwargs["master_uri"]
+
+        self._io_loop = kwargs.get("io_loop")
+        if self._io_loop is None:
+            self._io_loop = tornado.ioloop.IOLoop.current()
+
+        if self.crypt != "clear":
+            # we don't need to worry about auth as a kwarg, since its a singleton
+            self.auth = salt.crypt.AsyncAuth(self.opts, io_loop=self._io_loop)
+        log.debug(
+            "Connecting the Minion to the Master URI (for the return server): %s",
+            self.master_uri,
+        )
+        self.message_client = AsyncReqMessageClientPool(
+            self.opts,
+            args=(self.opts, self.master_uri,),
+            kwargs={"io_loop": self._io_loop},
+        )
+
+        self._closing = False
+
+    def close(self):
+        """
+        Since the message_client creates sockets and assigns them to the IOLoop we have to
+        specifically destroy them, since we aren't the only ones with references to the FDs
+        """
+        if self._closing:
+            return
+
+        if self._refcount > 1:
+            # Decrease refcount
+            with self._refcount_lock:
+                self._refcount -= 1
+            log.debug(
+                "This is not the last %s instance. Not closing yet.",
+                self.__class__.__name__,
+            )
+            return
+
+        log.debug("Closing %s instance", self.__class__.__name__)
+        self._closing = True
+        if hasattr(self, "message_client"):
+            self.message_client.close()
+
+        # Remove the entry from the instance map so that a closed entry may not
+        # be reused.
+        # This forces this operation even if the reference count of the entry
+        # has not yet gone to zero.
+        if self._io_loop in self.__class__.instance_map:
+            loop_instance_map = self.__class__.instance_map[self._io_loop]
+            if self._instance_key in loop_instance_map:
+                del loop_instance_map[self._instance_key]
+            if not loop_instance_map:
+                del self.__class__.instance_map[self._io_loop]
+
+    # pylint: disable=W1701
+    def __del__(self):
+        with self._refcount_lock:
+            # Make sure we actually close no matter if something
+            # went wrong with our ref counting
+            self._refcount = 1
+        try:
+            self.close()
+        except OSError as exc:
+            if exc.errno != errno.EBADF:
+                # If its not a bad file descriptor error, raise
+                raise
+
+    # pylint: enable=W1701
```

Comment:
**note** it's preferred to use the named pylint rather than the code. Makes it easier to know what warnings we're skipping.

Reply:
I'll fix it in the follow-up commit. This was carried over from zeromq.py implementation.